### PR TITLE
AutoBackNavigation Functionality

### PR DESCRIPTION
**Pull request by JitendraSaini1995**
when user not interacts with screen then it automatically navigates back after 10 sec.if user taps on the screen before 10 seconds then it will wait for another 10 seconds then after will back to previous screen.
[rec1.webm](https://github.com/sublimitysoftwares/OnTime_Android/assets/131166947/d9cd807f-9a70-4c5a-bc79-83d99cc225b0)
[rec2.webm](https://github.com/sublimitysoftwares/OnTime_Android/assets/131166947/5a190f3f-0848-4548-9332-3ee67b6aa54a)

### CODE REVIEW

File: app/src/main/java/com/allocate/ontime/business_logic/viewmodel/admin/AdminViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    ViewModel() {
+    val navigationFlows = autoBackNavigationManager.navigationFlow
+
+    init {
+        startInteraction()
+    }
+
+    fun startInteraction() {
+        autoBackNavigationManager.addOrUpdateScreens(OnTimeScreens.AdminScreen.name)
+    }
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/business_logic/viewmodel/admin/AdminViewModel.kt:22]: create one more method here which will remove this screen see  and understand the below code.
    fun resetAutoBack() {
        autoBackNavigationManager. removeScreens(OnTimeScreens.AdminScreen.name)
    }

File: app/src/main/java/com/allocate/ontime/presentation_logic/screens/admin/AdminScreen.kt [hunk 2]
```diff
@@ -56,9 +66,24 @@ fun AdminScreen(backToHome: (HomeScreenRoot) -> Unit) {
         })
     }
 
+    val hasNoUserInteractionAdminScreen = adminViewModel.navigationFlows.collectAsState()
+
+    if (hasNoUserInteractionAdminScreen.value) {
+        backToHome(HomeScreenRoot.HomeScreen)
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/presentation_logic/screens/admin/AdminScreen.kt:72]: adminViewModel. resetAutoBack(), Call this method from here, and do the same in other places as well

File: app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    private fun checkScreenInteractions(screenName: String) {
+        val currentTime = System.currentTimeMillis()
+        if (navigateScreens.isNotEmpty()) {
+            val screenLastTime = navigateScreens[screenName]
+            screenLastTime?.let { screenLastTimes ->
+                if (currentTime - screenLastTimes >= TIMEOUT_DURATION_MS) {
+                    CoroutineScope(Dispatchers.Main).launch {
+                        navigationFlowInternal.value = true
+                    }
+                    removeScreens(screenName)
```
Comment (mohit2988) [inline app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt:55]: remove it

File: app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    private val TIMEOUT_DURATION_MS = 10000L
+
+    fun addOrUpdateScreens(screen: String) {
+        navigateScreens[screen] = System.currentTimeMillis()
+        startTimeout(screen)
+    }
+
+    private fun removeScreens(screen: String) {
+        navigateScreens.remove(screen)
+        CoroutineScope(Dispatchers.Main).launch {
```
[the remark this answers is not in the document]
Comment (mohit2988) [reply app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt:30]: fun removeScreens(screen: String) {
        navigateScreens.remove(screen)
         navigationFlowInternal.value = false
         }
         
  This method will have only these 2 things inside, make it public.
  
  delay will lead some problem

File: app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class AutoBackNavigationManager @Inject constructor() {
+
+    private val navigateScreens: HashMap<String, Long> = hashMapOf()
+    val navigationFlowInternal = MutableStateFlow(false)
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt:18]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/viewmodel/home/HomeViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+import com.allocate.ontime.business_logic.autoback_navigation_manager.AutoBackNavigationManager
+import androidx.lifecycle.ViewModel
+import dagger.hilt.android.lifecycle.HiltViewModel
+import javax.inject.Inject
+
+@HiltViewModel
+class HomeViewModel @Inject constructor(autoBackNavigationManager: AutoBackNavigationManager) :
+    ViewModel() {
+    init {
+        autoBackNavigationManager.navigationFlowInternal.value = false
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/viewmodel/home/HomeViewModel.kt:12]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        }
+    }
+
+    private fun checkScreenInteractions(screenName: String) {
+        val currentTime = System.currentTimeMillis()
+        if (navigateScreens.isNotEmpty()) {
+            val screenLastTime = navigateScreens[screenName]
+            screenLastTime?.let { screenLastTimes ->
+                if (currentTime - screenLastTimes >= TIMEOUT_DURATION_MS) {
+                    CoroutineScope(Dispatchers.Main).launch {
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt:52]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+                        navigationFlowInternal.value = true
+                    }
+                    removeScreens(screenName)
+                }
+            }
+        } else {
+            CoroutineScope(Dispatchers.Main).launch {
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt:59]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+                navigationFlowInternal.value = false
+            }
+            stopTimeout()
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt:62]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt [hunk 10]
```diff
@@ -16,15 +20,33 @@ import retrofit2.Retrofit
...
     fun provideOnTimeRepository(api: OnTimeApi) =
         OnTimeRepository(api)
 
+    @Provides
+    fun provideSplashViewModel(repository: OnTimeRepository, daoRepository: DaoRepository) =
+        SplashViewModel(repository, daoRepository)
+
+    @Provides
+    @Singleton
+    fun provideAutoBackNavigationManager(): AutoBackNavigationManager {
```
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/di/AppModule.kt:38]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/viewmodel/admin/AdminViewModel.kt (hunk 1 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/viewmodel/admin/AdminViewModel.kt:22]: done!

File: app/src/main/java/com/allocate/ontime/presentation_logic/screens/admin/AdminScreen.kt (hunk 2 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/presentation_logic/screens/admin/AdminScreen.kt:72]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt (hunk 3 above)
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt:55]: done!

File: app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (JitendraSaini1995) [reply app/src/main/java/com/allocate/ontime/business_logic/autoback_navigation_manager/AutoBackNavigationManager.kt:30]: done!

**Review by mohit2988 — APPROVED** [review]
approve

**Pull request merged**